keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -484,26 +493,158 @@ while another might validate the containers declared in the `podTemplate` of a
+
+Alternative considered: `PolicyConfiguration` kind with config embedded
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyConfiguration
+metadata:
+  name: "replica-limit-prod.example.com"
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: NotIn,
+      values: ["test"]
+  config:
+    apiVersion: rules.example.com/v1
+    kind: ReplicaLimit
+    spec:
+      maxReplicas: 100
+```
+
+Pros:
+
+- Matching criteria is fully defined and validated in a builtin type
+
+Cons:
+
+- Embedded config needs the same treatment as custom resource, but
+  reimplementing it all on an embedded resource is, at best, highly impractical
+  in the apiserver as it exists today
+
+Alternative: Generate CRDs
+
+This allows the apiserver to combine the configuration schema with the match
+schema and then generate a CRD.  The apiserver then can run a control loop
+to always keep the CRD match stanzas in sync with what is expected.
+
+Pros:
+
+- Policy author doesn't need to define the CRD. It can instead be generated for
+  them.
 
 Cons:
 
 - This implies that the policy owns the configuration. But we have use cases
   where multiple policies share the same configuration. For this model it seems
-  misleading and potentially to have both policies defining the configuration.
+  misleading and potentially problematic to have both policies attempting to
+  define the same configuration.
 
 ##### Match Criteria
 
 During admission, the Kubernetes API server will validate the resource being
 admitted against all policy configuration resources that match the resource.
 
+While webhook match rules give a good sense of what types of capabilities might
+be needed, they serve a slightly different purpose.  Webhook match rules make it
+possible to avoid webhook requests, which incur latency and impact availability,
+for resources that don't need to be evaluated by the webhook. CEL expressions
+have a comparatively small impact of latency and are in-process (and so do not
+have the same impact to availability).
+
+For CEL expressions, the primary benefits of match criteria are:
+
+- Match criteria establishes bounds on what sort of admission requests the CEL
+  expressions must consider. The CEL expressions can be written knowing that the
+  match criteria has filtered out requests that is does not need to consider.
+- Match criteria is available on policy configurations and allows the
+  configuration author to further constrain what resources the particular
+  configuration applies to.
+
+Matching is performed in quite a few systems across Kubernetes:
+
+| Match type                               | Usages in existing matchers  | Support?                       |
+| ---------------------------------------- | ---------------------------- | ------------------------------ |
+| namespace selectors                      | WH/Audit/P&F                 | phase 1                        |
```

Comment:
some G8r users have leveraged prefix/suffix matching for namespace/name

Reply:
For those use cases, would doing the prefix check in CEL be acceptable? I.e. use wider match criteria and then have the CEL expression allow resources where the namespace/name don't match the prefix/suffix?

Reply:
I've tracked this on the KEP with a UNRESOLVED and a link back to this thread.